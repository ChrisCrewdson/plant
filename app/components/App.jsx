const React = require('react');
const actions = require('../actions');
const PropTypes = require('prop-types');

class App extends React.Component {
  static contextTypes = {
    store: PropTypes.object.isRequired,
  };

  componentWillMount() {
    const { store } = this.context;
    const users = store.getState().get('users');
    if (!users || users.size === 0) {
      store.dispatch(actions.loadUsersRequest());
    }

    const locations = store.getState().get('locations');
    if (!locations || locations.size === 0) {
      store.dispatch(actions.loadLocationsRequest());
    }

    // const user = store.getState().get('user');
    // if(user && user.get('isLoggedIn', false)) {
    //   const locationIds = user.get('locationIds');
    //   if(!locationIds) {
    //     const userId = user.get('_id');
    //     store.dispatch(actions.loadLocationsRequest(userId));
    //   }
    // }
  }

  render() {
    return (
      <div className="react-root">
        {this.props.children}
      </div>
    );
  }
}

App.propTypes = {
  // eslint-disable-next-line react/forbid-prop-types
  children: PropTypes.object.isRequired,
};

module.exports = App;
