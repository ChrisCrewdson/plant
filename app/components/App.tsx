import React from 'react';
import PropTypes from 'prop-types';
import { actionFunc } from '../actions';

export default class App extends React.Component {
  // TODO: When tsc 3.7+ is in use remove the ! to see hint text on how to change this.
  context!: PlantContext;

  static propTypes = {
    // eslint-disable-next-line react/forbid-prop-types
    children: PropTypes.object.isRequired,
  };

  static contextTypes = {
    // eslint-disable-next-line react/forbid-prop-types
    store: PropTypes.object.isRequired,
  };

  // eslint-disable-next-line camelcase
  UNSAFE_componentWillMount() {
    const { store } = this.context;
    const { users = {}, locations = {} } = store.getState();

    // TODO: This will cause a problem for a non-initialized site
    // that has zero users and zero locations as these values
    // will be 0 which is falsy.
    const usersCount = Object.keys(users).length;
    const locationsCount = Object.keys(locations).length;

    if (!usersCount) {
      store.dispatch(actionFunc.loadUsersRequest());
    }

    if (!locationsCount) {
      store.dispatch(actionFunc.loadLocationsRequest());
    }
  }

  render() {
    const { children } = this.props;
    return (
      <div className="react-root">
        {children}
      </div>
    );
  }
}
