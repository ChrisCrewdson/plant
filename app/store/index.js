const { createStore, applyMiddleware } = require('redux');
const reducers = require('../reducers'); // combineReducers already called on reducers in her)e
const api = require('../middleware/api');
const { setupSubscribe: userSubscribe } = require('./user');
const Immutable = require('immutable');

let middleware = [api];

if (process.env.NODE_ENV !== 'production') {
  const logger = require('../middleware/logger');
  middleware = [logger, api];
}

// Add the api to the pipeline/chain
const createStoreWithMiddleware = applyMiddleware(...middleware)(createStore);

const store = createStoreWithMiddleware(reducers, Immutable.fromJS(window.__INITIAL_STATE__ || {}));

userSubscribe(store);

module.exports = store;
