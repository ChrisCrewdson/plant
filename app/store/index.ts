export {}; // To get around: Cannot redeclare block-scoped variable .ts(2451)

const { createStore, applyMiddleware } = require('redux');

const seamless = require('seamless-immutable').static;
const reducers = require('../reducers'); // combineReducers already called on reducers in her)e
const api = require('../middleware/api');
const { setupSubscribe: userSubscribe } = require('./user');

const middleware = [api];

if (process.env.NODE_ENV !== 'production') {
  // eslint-disable-next-line global-require
  const logger = require('../middleware/logger');
  middleware.unshift(logger);
}

// Add the api to the pipeline/chain
const createStoreWithMiddleware = applyMiddleware(...middleware)(createStore);

// @ts-ignore - __INITIAL_STATE__ is added on the server
const initialState = seamless.from(window.__INITIAL_STATE__ || {});
const store = createStoreWithMiddleware(reducers, initialState);

userSubscribe(store);

module.exports = store;
