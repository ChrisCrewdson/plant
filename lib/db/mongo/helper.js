const _ = require('lodash');

module.exports = class Helper {
  static removeEmpty(doc) {
    if (!doc) {
      return doc;
    }
    if (_.isArray(doc)) {
      return doc.map(item => Helper.removeEmpty(item));
    }
    return _.pickBy(doc, item => item !== '');
  }

  /**
   * Helper function to convert _id from MongoId to string. Used in reads
   * @param {object} obj - Object that might have an _id
   * @returns {object} - the same object with a converted _id
   */
  static convertIdToString(obj) {
    if (_.isArray(obj)) {
      return obj.map(Helper.convertIdToString);
    }
    if (obj && obj._id) {
      // eslint-disable-next-line no-param-reassign
      obj._id = obj._id.toString();
    }
    return obj;
  }
};
