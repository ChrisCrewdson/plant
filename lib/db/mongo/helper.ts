import _ from 'lodash';

export class Helper {
  /**
   * Removes empty string values from properties
   * @param doc - an object with props that might have empty string values
   */
  static removeEmpty(doc: object | Array<any>): object | Array<any> {
    if (!doc) {
      return doc;
    }
    if (_.isArray(doc)) {
      return (doc as Array<any>).map((item) => Helper.removeEmpty(item));
    }
    return _.pickBy(doc, (item: any) => item !== '');
  }

  /**
   * Helper function to convert _id from MongoId to string. Used in reads
   * @param obj - Object that might have an _id
   * @returns - the same object with a converted _id
   */
  static convertIdToString(obj: any): object {
    if (_.isArray(obj)) {
      return (obj as Array<any>).map(Helper.convertIdToString);
    }
    if (obj && obj._id) {
      // eslint-disable-next-line no-param-reassign
      obj._id = obj._id.toString();
    }
    return obj;
  }
}
