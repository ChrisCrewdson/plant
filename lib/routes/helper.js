const logger = require('../logging/logger').create('routes-helper');

function helper (app) {

  app.use((req, res, next) => {
    logger.trace('hit here in helper');
    if(req && req.headers && req.headers['user-agent']) {
      if(req.headers['user-agent'].toLowerCase().includes('facebookexternalhit')) {
        const {headers, originalUrl, query} = req;
        logger.info('facebookexternalhit', {headers, originalUrl, query});
      }
    }
    next();
  });

};

module.exports = helper;
