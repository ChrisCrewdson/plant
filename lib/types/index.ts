
// Definitions file for Plant Project

// Currently NonEmptyList is not used anywhere.
// Plan to tag the plantIds in a Note with this as there
// always has to be at least 1 plant for each note.
type NonEmptyList<T> = T[] & { 0: T };

// Now included with Typescript 3.5+
// type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>

/*
In TypeScript, just as in ECMAScript 2015, any file containing a top-level import or export
is considered a module. Conversely, a file without any top-level import or export declarations
is treated as a script whose contents are available in the global scope (and therefore to
modules as well).
*/

type GeoType = 'Point';

interface GeoCoords {
  /** Longitude */
  0: number;
  /** Latitude */
  1: number;
}

interface Geo {
  type: GeoType; // "Point" - is this the only value I know of that's valid
  coordinates: GeoCoords;
}
interface Window {
  [key: string]: any;
  // FormData: Function;
}

type Mutable<T> = {
  -readonly [P in keyof T]: T[P];
};

interface Global extends NodeJS.Global {
  [key: string]: any;
  // window: Window;
  // requestAnimationFrame?: (callback: Function) => void;
}

type Html5InputTypes =
  | 'button'
  | 'checkbox'
  | 'color'
  | 'date'
  | 'datetime-local'
  | 'email'
  | 'file'
  | 'hidden'
  | 'image'
  | 'month'
  | 'number'
  | 'password'
  | 'radio'
  | 'range'
  | 'reset'
  | 'search'
  | 'submit'
  | 'tel'
  | 'text'
  | 'time'
  | 'url'
  | 'week';
