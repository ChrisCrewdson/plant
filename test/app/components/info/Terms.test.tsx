import React from 'react';
import renderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import { Dispatch } from 'redux';

import { MuiThemeProvider } from '@material-ui/core/styles';

import Terms from '../../../../app/components/info/Terms';
import store from '../../../../app/store';
import App from '../../../../app/components/App';
import { PlantAction } from '../../../../lib/types/redux-payloads';
import { theme } from '../../../helper';

describe('Terms', () => {
  const storeDispatch = store.dispatch;

  beforeAll(() => {
    store.dispatch = (() => {}) as Dispatch<PlantAction<any>>;
  });

  afterAll(() => {
    store.dispatch = storeDispatch;
  });

  test('Terms should be rendered', () => {
    const component = renderer.create(
      <MuiThemeProvider theme={theme}>
        <Provider store={store}>
          <App>
            <MemoryRouter>
              <Terms />
            </MemoryRouter>
          </App>
        </Provider>
      </MuiThemeProvider>);
    const badgeTotals = component.toJSON();
    expect(badgeTotals).toMatchSnapshot();
  });
});
