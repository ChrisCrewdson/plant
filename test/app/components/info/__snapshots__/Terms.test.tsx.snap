// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Terms Terms should be rendered 1`] = `
<div
  className="react-root"
>
  <div
    className="page"
  >
    <div
      className="makeStyles-root-1"
    >
      <header
        className="MuiPaper-root MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorPrimary MuiPaper-elevation4"
      >
        <div
          className="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
        >
          <h4
            className="MuiTypography-root makeStyles-title-3 MuiTypography-h4"
          >
            <a
              href="/"
              onClick={[Function]}
              style={
                Object {
                  "marginRight": "10px",
                }
              }
            >
              Plant
            </a>
          </h4>
          <h4
            className="MuiTypography-root MuiTypography-h4"
          >
            <a
              href="/login"
              onClick={[Function]}
              style={
                Object {
                  "marginRight": "10px",
                }
              }
            >
              Login
            </a>
            <a
              href="/help"
              onClick={[Function]}
              title="help"
            >
              Help
            </a>
          </h4>
        </div>
      </header>
    </div>
    <div
      id="main"
    >
      <div
        className="MuiPaper-root MuiPaper-elevation24 MuiPaper-rounded"
        style={
          Object {
            "display": "inline-block",
            "margin": 20,
            "padding": 20,
            "width": "100%",
          }
        }
      >
        <div
          dangerouslySetInnerHTML={
            Object {
              "__html": "<h1 id=\\"introduction\\">Introduction</h1>
<p>These terms and conditions govern your use of this website; by using this website, you accept these terms and conditions in full. If you disagree with these terms and conditions or any part of these terms and conditions, you must not use this website. </p>
<h1 id=\\"license-to-use-website\\">License to use website</h1>
<p>Unless otherwise stated, Plaaant and/or its licensors and/or its contributors own the intellectual property rights in the website and material on the website. Subject to the license below, all these intellectual property rights are reserved.</p>
<p>You may view, download for caching purposes only, and print pages from the website for your own personal use, subject to the restrictions set out below and elsewhere in these terms and conditions.  </p>
<p>You must not:</p>
<ul>
<li>Republish material from this website (including republication on another website);</li>
<li>Sell, rent or sub-license material from the website;</li>
<li>Show any material from the website in public;</li>
<li>Reproduce, duplicate, copy or otherwise exploit material on this website for a commercial purpose;</li>
<li>Edit or otherwise modify any material on the website unless it is &quot;your user content&quot;;</li>
<li>Redistribute material from this website.</li>
</ul>
<h1 id=\\"acceptable-use\\">Acceptable use</h1>
<p>You must not use this website in any way that causes, or may cause, damage to the website or impairment of the availability or accessibility of the website; or in any way which is unlawful, illegal, fraudulent or harmful, or in connection with any unlawful, illegal, fraudulent or harmful purpose or activity.</p>
<p>You must not use this website to copy, store, host, transmit, send, use, publish or distribute any material which consists of (or is linked to) any spyware, computer virus, Trojan horse, worm, keystroke logger, rootkit or other malicious computer software.</p>
<p>You must not conduct any systematic or automated data collection activities (including without limitation scraping, data mining, data extraction and data harvesting) on or in relation to this website without Plaaant&#39;s express written consent.</p>
<p>You must not use this website to transmit or send unsolicited commercial communications.</p>
<p>You must not use this website for any purposes related to marketing without Plaaant&#39;s express written consent.</p>
<h1 id=\\"user-content\\">User content</h1>
<p>In these terms and conditions, “your user content” means material (including without limitation text, images, audio material, video material and audio-visual material) that you submit to this website, for whatever purpose.</p>
<p>You grant to Plaaant a worldwide, irrevocable, non-exclusive, royalty-free license to use, reproduce, adapt, publish, translate and distribute your user content in any existing or future media.  You also grant to Plaaant the right to sub-license these rights, and the right to bring an action for infringement of these rights.</p>
<p>Your user content must not be illegal or unlawful, must not infringe any third party&#39;s legal rights, and must not be capable of giving rise to legal action whether against you or Plaaant or a third party (in each case under any applicable law).  </p>
<p>You must not submit any user content to the website that is or has ever been the subject of any threatened or actual legal proceedings or other similar complaint.</p>
<p>Plaaant reserves the right to edit or remove any material submitted to this website, or stored on Plaaant&#39;s servers, or hosted or published upon this website.</p>
<p>Notwithstanding Plaaant&#39;s rights under these terms and conditions in relation to user content, Plaaant does not undertake to monitor the submission of such content to, or the publication of such content on, this website.</p>
<h1 id=\\"no-warranties\\">No warranties</h1>
<p>This website is provided “as is” without any representations or warranties, express or implied.  Plaaant makes no representations or warranties in relation to this website or the information and materials provided on this website.  </p>
<p>Without prejudice to the generality of the foregoing paragraph, Plaaant does not warrant that:</p>
<ul>
<li>this website will be constantly available, or available at all; or</li>
<li>the information on this website is complete, true, accurate or non-misleading.</li>
</ul>
<p>Nothing on this website constitutes, or is meant to constitute, advice of any kind. If you require advice in relation to any legal, financial or medical matter you should consult an appropriate professional.</p>
<h1 id=\\"limitations-of-liability\\">Limitations of liability</h1>
<p>Plaaant will not be liable to you (whether under the law of contact, the law of torts or otherwise) in relation to the contents of, or use of, or otherwise in connection with, this website:</p>
<ul>
<li>to the extent that the website is provided free-of-charge, for any direct loss;</li>
<li>for any indirect, special or consequential loss; or</li>
<li>for any business losses, loss of revenue, income, profits or anticipated savings, loss of contracts or business relationships, loss of reputation or goodwill, or loss or corruption of information or data.</li>
</ul>
<p>These limitations of liability apply even if Plaaant has been expressly advised of the potential loss.</p>
<h1 id=\\"exceptions\\">Exceptions</h1>
<p>Nothing in this website disclaimer will exclude or limit any warranty implied by law that it would be unlawful to exclude or limit; and nothing in this website disclaimer will exclude or limit Plaaant&#39;s liability in respect of any:</p>
<ul>
<li>death or personal injury caused by Plaaant&#39;s negligence;</li>
<li>fraud or fraudulent misrepresentation on the part of Plaaant; or</li>
<li>matter which it would be illegal or unlawful for Plaaant to exclude or limit, or to attempt or purport to exclude or limit, its liability. </li>
</ul>
<h1 id=\\"reasonableness\\">Reasonableness</h1>
<p>By using this website, you agree that the exclusions and limitations of liability set out in this website disclaimer are reasonable.  </p>
<p>If you do not think they are reasonable, you must not use this website.</p>
<h1 id=\\"other-parties\\">Other parties</h1>
<p>You accept that, as a limited liability entity, Plaaant has an interest in limiting the personal liability of its officers and employees.  You agree that you will not bring any claim personally against Plaaant&#39;s officers or employees in respect of any losses you suffer in connection with the website.</p>
<p>Without prejudice to the foregoing paragraph, you agree that the limitations of warranties and liability set out in this website disclaimer will protect Plaaant&#39;s officers, employees, agents, subsidiaries, successors, assigns and sub-contractors as well as Plaaant. </p>
<h1 id=\\"unenforceable-provisions\\">Unenforceable provisions</h1>
<p>If any provision of this website disclaimer is, or is found to be, unenforceable under applicable law, that will not affect the enforceability of the other provisions of this website disclaimer.</p>
<h1 id=\\"indemnity\\">Indemnity</h1>
<p>You hereby indemnify Plaaant and undertake to keep Plaaant indemnified against any losses, damages, costs, liabilities and expenses (including without limitation legal expenses and any amounts paid by Plaaant to a third party in settlement of a claim or dispute on the advice of Plaaant&#39;s legal advisers) incurred or suffered by Plaaant arising out of any breach by you of any provision of these terms and conditions, or arising out of any claim that you have breached any provision of these terms and conditions.</p>
<h1 id=\\"breaches-of-these-terms-and-conditions\\">Breaches of these terms and conditions</h1>
<p>Without prejudice to Plaaant&#39;s other rights under these terms and conditions, if you breach these terms and conditions in any way, Plaaant may take such action as Plaaant deems appropriate to deal with the breach, including suspending your access to the website, prohibiting you from accessing the website, blocking computers using your IP address from accessing the website, contacting your internet service provider to request that they block your access to the website and/or bringing court proceedings against you.</p>
<h1 id=\\"variation\\">Variation</h1>
<p>Plaaant may revise these terms and conditions from time-to-time.  Revised terms and conditions will apply to the use of this website from the date of the publication of the revised terms and conditions on this website.  Please check this page regularly to ensure you are familiar with the current version.</p>
<h1 id=\\"assignment\\">Assignment</h1>
<p>Plaaant may transfer, sub-contract or otherwise deal with Plaaant&#39;s rights and/or obligations under these terms and conditions without notifying you or obtaining your consent.</p>
<p>You may not transfer, sub-contract or otherwise deal with your rights and/or obligations under these terms and conditions.  </p>
<h1 id=\\"severability\\">Severability</h1>
<p>If a provision of these terms and conditions is determined by any court or other competent authority to be unlawful and/or unenforceable, the other provisions will continue in effect.  If any unlawful and/or unenforceable provision would be lawful or enforceable if part of it were deleted, that part will be deemed to be deleted, and the rest of the provision will continue in effect. </p>
<h1 id=\\"entire-agreement\\">Entire agreement</h1>
<p>These terms and conditions constitute the entire agreement between you and Plaaant in relation to your use of this website, and supersede all previous agreements in respect of your use of this website.</p>
<h1 id=\\"law-and-jurisdiction\\">Law and jurisdiction</h1>
<p>These terms and conditions will be governed by and construed in accordance with United States Law, and any disputes relating to these terms and conditions will be subject to the exclusive jurisdiction of the courts of Arizona.</p>
",
            }
          }
        />
      </div>
    </div>
    <nav
      className="navbar navbar-default navbar-fixed-bottom"
    >
      <div
        className="container-fluid"
      >
        <a
          href="/"
          onClick={[Function]}
        >
          Home
        </a>
        <a
          href="/privacy"
          onClick={[Function]}
          style={
            Object {
              "marginLeft": "10px",
            }
          }
        >
          Privacy
        </a>
        <a
          href="/terms"
          onClick={[Function]}
          style={
            Object {
              "marginLeft": "10px",
            }
          }
        >
          Terms
        </a>
      </div>
    </nav>
  </div>
</div>
`;
